### PR TITLE
EnC support for source-generated files

### PR DESCRIPTION
**Pull request by tmat**
Builds on top of https://github.com/dotnet/roslyn/pull/51420.

Adds ability to detect changes in files generated by source generators: https://github.com/dotnet/roslyn/issues/50283.

Fixes an issue with merging partial type edits.

### CODE REVIEW

**Comment by tmat** [comment]
@davidwengier PTAL

**Review by davidwengier — APPROVED** [review]

File: src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs [hunk 1]
```diff
@@ -25,9 +25,9 @@ internal abstract class DefinitionMap
             public readonly IMethodSymbolInternal PreviousMethod;
             public readonly Func<SyntaxNode, SyntaxNode> SyntaxMap;
 
-            public MappedMethod(IMethodSymbolInternal previousMethodOpt, Func<SyntaxNode, SyntaxNode> syntaxMap)
+            public MappedMethod(IMethodSymbolInternal previousMethod, Func<SyntaxNode, SyntaxNode> syntaxMap)
             {
-                PreviousMethod = previousMethodOpt;
+                PreviousMethod = previousMethod;
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/Compilers/Core/Portable/Emit/EditAndContinue/DefinitionMap.cs:30]: I'll follow up with annotating this file.

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs [hunk 2]
```diff
@@ -191,22 +191,16 @@ private static Project CreateProject(ProjectId projectId, Solution solution)
...
         /// Gets a document or a source generated document in this solution with the specified document ID.
         /// </summary>
-        internal async ValueTask<Document?> GetDocumentAsync(DocumentId? documentId, bool includeSourceGenerated = false, CancellationToken cancellationToken = default)
+        internal ValueTask<Document?> GetDocumentAsync(DocumentId? documentId, bool includeSourceGenerated = false, CancellationToken cancellationToken = default)
         {
             var project = GetProject(documentId?.ProjectId);
             if (project == null)
             {
-                return null;
+                return default;
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:199]: That's actually safe?

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs [hunk 3]
```diff
@@ -271,12 +285,15 @@ internal async ValueTask<IEnumerable<Document>> GetAllRegularAndSourceGeneratedD
...
             if (generatedDocumentState != null)
             {
-                return ImmutableHashMapExtensions.GetOrAdd(ref _idToSourceGeneratedDocumentMap, generatedDocumentState.Id, s_createSourceGeneratedDocumentFunction, (generatedDocumentState, this));
+                return GetOrCreateSourceGeneratedDocument(generatedDocumentState);
             }
 
             return null;
         }
 
+        internal SourceGeneratedDocument GetOrCreateSourceGeneratedDocument(SourceGeneratedDocumentState state)
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:294]: Was this supposed to get called somewhere else too?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 4]
```diff
@@ -342,16 +341,62 @@ private static async Task PopulateChangedAndAddedDocumentsAsync(CommittedSolutio
...
             }
+
+            // TODO: support document removal/rename (see https://github.com/dotnet/roslyn/issues/41144, https://github.com/dotnet/roslyn/issues/49013).
+
+            // Given the following assumptions:
+            // - source generators are deterministic,
+            // - source documents, metadata references and compilation options have not changed,
+            // - additional documents have not changed,
+            // - analyzer config documents have not changed,
+            // the outputs of source generators will not change.
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:362]: So what's the intent here for project-to-project references? That you're not running? I don't disagree with the decision, I wasn't sure if there were some other cases you have in mind.

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs [hunk 5]
```diff
@@ -136,55 +136,73 @@ public async Task<(Document? Document, DocumentState State)> GetDocumentAndState
...
+                            return (sourceGeneratedDocument, DocumentState.MatchesBuildOutput);
+                        }
 
-                        case DocumentState.None:
-                            throw ExceptionUtilities.Unreachable;
+                        // The current document is source-generated therefore the corresponding one is not present in the base solution.
+                        if (currentDocument is SourceGeneratedDocument)
+                        {
+                            return (null, DocumentState.MatchesBuildOutput);
+                        }
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs:188]: This the case where the generator is now producing one more generated file than it didn't before?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs [hunk 6]
```diff
@@ -342,16 +341,62 @@ private static async Task PopulateChangedAndAddedDocumentsAsync(CommittedSolutio
...
+            // 
+            // Currently it's not possible to change compilation options (Project System is readonly during debugging).
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/EditAndContinue/EditSession.cs:364]: I wonder what happens if we use our code fixes that change compilation options. Not sure how much I want to try though!

File: src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:199]: I think so. Why wouldn't be?

Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:199]: https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/System.Threading.Tasks.Extensions/tests/ValueTaskTests.cs#L36

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/Project.cs:294]: It is: https://github.com/dotnet/roslyn/blob/af7a90f6404d8a34b8e58ede5210268ad41b6b31/src/Features/Core/Portable/EditAndContinue/EditSession.cs#L387

File: src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs (hunk 5 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/CommittedSolution.cs:188]: Yes. It's a newly added source-generated document.

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs (hunk 4 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs:362]: You mean the generator depending on content of other projects to which it gets thru references of the compilation?

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs:362]: How does the IDE deal with that for other services, like completion?

File: src/Features/Core/Portable/EditAndContinue/EditSession.cs (hunk 6 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/EditSession.cs:364]: Probably nothing good if the options affect emit. We'll need to handle that some day, but not today :)

**Pull request merged**